### PR TITLE
Keep const-tensor dim order canonical when a dimension has size 1

### LINKED ISSUE #22520: Depthwise separable convolution computation by portable kernels expects wrong dim_order
### 🐛 Describe the bug

Running `DSCNNKWS` quantized executorch program in channel last using portable kernels produces the error below. The error indicates that a convolution (probably depthwise separable) expects a different dim order of the input tensor.
The error was discovered in a newly added NXP test for MLPerf Tiny Keyword Spotting in this PR:
https://github.com/pytorch/executorch/pull/22519
After investigating further, I believe the issue stems from `legalize_portable_dim_order_passes`, however I was unable to resolve it.
The error also looks quite similar to one of my previously reported issues:
https://github.com/pytorch/executorch/issues/20804

```
CRITICAL executorch.backends.nxp.tests.nsys_testing:nsys_testing.py:766 D 00:00:00.000010 executorch:operator_registry.cpp:116] Successfully registered all kernels from shared library: NOT_SUPPORTED
D 00:00:00.000052 executorch:operator_registry.cpp:116] Successfully registered all kernels from shared library: NOT_SUPPORTED
D 00:00:00.000162 executorch:operator_registry.cpp:116] Successfully registered all kernels from shared library: NOT_SUPPORTED
D 00:00:00.000736 executorch:method.cpp:864] Loading method: forward.
Loading file /home/nxg16998/code/executorch-integration/.outputs/test_mlperf_tiny_kws_mse_cpu_vs_npu[True]/dataset/test/example_Left_2_59_i00.bin
D 00:00:00.001189 executorch:method.cpp:1717] Executing method: forward.
E 00:00:00.003222 executorch:tensor_util_portable.cpp:64] Expected tensor to have default or channels last dim order, but got
E 00:00:00.003271 executorch:tensor_util_portable.cpp:68]     dim_order(0): 0
E 00:00:00.003276 executorch:tensor_util_portable.cpp:68]     dim_order(1): 2
E 00:00:00.003277 executorch:tensor_util_portable.cpp:68]     dim_order(2): 1
E 00:00:00.003278 executorch:tensor_util_portable.cpp:68]     dim_order(3): 3
E 00:00:00.003280 executorch:kernel_ops_util.cpp:384] Check failed (tensor_is_default_or_channels_last_dim_order(in)): 
E 00:00:00.003282 executorch:op_convolution.cpp:363] Check failed (check_convolution_args( in, weight, bias, stride, padding, dilation, transposed, output_padding, groups, out)): 
E 00:00:00.004056 executorch:method.cpp:1483] KernelCall failed at instruction 0:13 in operator aten::convolution.out: 0x12
E 00:00:00.004059 executorch:method.cpp:1493] arg 0 with type id 1
E 00:00:00.004060 executorch:method.cpp:1493] arg 1 with type id 1
E 00:00:00.004060 executorch:method.cpp:1493] arg 2 with type id 1
E 00:00:00.004061 executorch:method.cpp:1493] arg 3 with type id 8
E 00:00:00.004062 executorch:method.cpp:1493] arg 4 with type id 8
E 00:00:00.004062 executorch:method.cpp:1493] arg 5 with type id 8
E 00:00:00.004063 executorch:method.cpp:1493] arg 6 with type id 5
E 00:00:00.004063 executorch:method.cpp:1493] arg 7 with type id 8
E 00:00:00.004064 executorch:method.cpp:1493] arg 8 with type id 4
E 00:00:00.004064 executorch:method.cpp:1493] arg 9 with type id 1
E 00:00:00.004065 executorch:method.cpp:1493] arg 10 with type id 1
Execution of method forward failed with status 18...
```

### Versions

Collecting environment information...
PyTorch version: 2.13.0+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.31.10
Libc version: glibc-2.39

Python version: 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Address sizes:                        46 bits physical, 48 bits virtual
Byte Order:                           Little Endian
CPU(s):                               16
On-line CPU(s) list:                  0-15
Vendor ID:                            GenuineIntel
Model name:                           13th Gen Intel(R) Core(TM) i5-1350P
CPU family:                           6
Model:                                186
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Stepping:                             2
BogoMIPS:                             4377.61
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni vnmi umip waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize flush_l1d arch_capabilities
Virtualization:                       VT-x
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            384 KiB (8 instances)
L1i cache:                            256 KiB (8 instances)
L2 cache:                             10 MiB (8 instances)
L3 cache:                             12 MiB (1 instance)
NUMA node(s):                         1
NUMA node0 CPU(s):                    0-15
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Mitigation; Clear Register File
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected

Versions of relevant libraries:
[pip3] executorch==1.5.0+9071414
[pip3] flake8==6.1.0
[pip3] flake8-breakpoint==1.1.0
[pip3] flake8-bugbear==24.4.26
[pip3] flake8-comprehensions==3.14.0
[pip3] flake8-plugin-utils==1.3.3
[pip3] flake8-pyi==23.5.0
[pip3] mypy==1.14.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.0.0
[pip3] optree==0.19.1
[pip3] pytorch_tokenizers==1.3.0
[pip3] torch==2.13.0+cpu
[pip3] torchao==0.18.0.dev20260729+cpu
[pip3] torchaudio==2.11.0+cpu
[pip3] torchdata==0.11.0+cpu
[pip3] torchsr==1.0.4
[pip3] torchtune==0.0.0
[pip3] torchvision==0.28.0+cpu
[conda] Could not collect

cc @larryliu0820 @manuelcandales @JakeStevens @JacobSzwejbka @lucylq @robert-kalmar @digantdesai @rascani

### PR DESCRIPTION
**Pull request by akashjainn**
Fixes #22520.

The tensor the portable kernel rejects is the weight, not the activation:

```
Expected tensor to have default or channels last dim order, but got
  dim_order(0): 0  dim_order(1): 2  dim_order(2): 1  dim_order(3): 3
[kernel_ops_util.cpp:386] Check failed (tensor_is_default_or_channels_last_dim_order(weight))
```

A depthwise weight is (C_out, 1, kH, kW). In channels-last that's strides like (9, 1, 3, 1), where dims 1 and 3 tie at stride 1 only because the input-channel dim has size 1. dim_order_from_stride sees strides alone and its stable descending sort keeps dim 1 ahead of dim 3:

```python
>>> w = torch.randn(8, 1, 3, 3).to(memory_format=torch.channels_last)
>>> w.stride()
(9, 1, 3, 1)
>>> w.dim_order()                        # torch
(0, 2, 3, 1)                             # channels-last, accepted
>>> dim_order_from_stride(w.stride())    # executorch
(0, 2, 1, 3)                             # neither, rejected
```

For (8, 2, 3, 3) both agree on (0, 2, 3, 1) - you need the size-1 dim to see the disagreement.

The emitter had already decided the tensor was fine, leaving it alone precisely because is_contiguous() or is_contiguous(channels_last) held, and then contradicted itself a line later by deriving the dim order from strides:

```python
if not (real_tensor.is_contiguous() or real_tensor.is_contiguous(memory_format=torch.channels_last)):
    real_tensor = real_tensor.contiguous()
...
spec.dim_order = dim_order_from_stride(spec.stride)   # can now say "neither"
```

So take the order from torch.Tensor.dim_order(), which resolves the tie against the tensor's sizes, and fall back to the stride-only computation for tensors it can't answer for, like ones with unbacked symbolic strides. Applied at the two places that hold a real tensor: TensorSpec.from_tensor and the emitter's constant path.

It isn't only depthwise - any weight with a size-1 input-channel dim hits it, including a plain single-channel conv. Over 24 convolutions (channels 1/4/8, dense and depthwise, 1x1 and 3x3, contiguous and channels-last), 6 fail before this change and all 24 pass after. Every failure is channels-last with a size-1 channel dim.

Repro needs no quantization and no backend:

```python
m = torch.nn.Conv2d(8, 8, 3, padding=1, groups=8).eval().to(memory_format=torch.channels_last)
x = torch.randn(1, 8, 8, 8).to(memory_format=torch.channels_last)
prog = to_edge_transform_and_lower(export(m, (x,), strict=True), partitioner=[]).to_executorch()
_load_for_executorch_from_buffer(prog.buffer).run_method("forward", (x,))   # 0x12 before this change
```

Testing: new exir/tests/test_channels_last_depthwise_conv.py, 3 tests plus 24 subtests, and 8 of them fail without the fix. New unit tests in exir/tests/test_tensor.py including the symbolic-stride fallback. test_tensor.py 13 passed, exir/emit/test/ 69 passed, test_dim_order_utils.py and test_legalize_portable_dim_order_pass.py 20 passed. flake8 and ufmt clean.

The end-to-end test compares runtime output against eager rather than only checking it runs, so a silently wrong layout would fail too.

One limitation worth stating: I ran these against an installed ExecuTorch wheel with the two changed modules overlaid, because the source tree isn't importable without a full build on my machine. A few repo tests (test_memory_planning.py) reference symbols newer than the wheel and couldn't be collected that way, so CI is the real check for those.

On #21621 - it reports the same dim_order(0, 2, 1, 3) signature for yolo26 on the portable backend, and that model uses depthwise convolutions, so it's plausibly the same cause. I couldn't run yolo26 to confirm, so I haven't marked it fixed.

cc @JacobSzwejbka @angelayi @larryliu0820 @lucylq

### CODE REVIEW

**Comment by jathu** [comment]
Closing due to purely being AI generated without human involvement.
